extension/runtime/src/main/java/io/quarkiverse/infinispan/embedded/runtime/cache/InfinispanCacheImpl.java

```diff
@@ -0,0 +1,220 @@
+                                    resultAsync.complete(decodeNull(existing));
+                                }
+                                computationResults.remove(key);
+                            });
+                    return resultAsync;
+                }));
+    }
+
+    @Override
+    public <K, V> Uni<V> getAsync(K key, Function<K, Uni<V>> valueLoader) {
+        Context context = Vertx.currentContext();
+
+        return Uni.createFrom().completionStage(CompletionStages.handleAndCompose(cache.getAsync(key), (v1, ex1) -> {
+            if (ex1 != null) {
+                return CompletableFuture.failedFuture(ex1);
+            }
+
+            if (v1 != null) {
+                return CompletableFuture.completedFuture(decodeNull(v1));
+            }
+
+            CompletableFuture<V> resultAsync = new CompletableFuture<>();
+            CompletableFuture<V> computedValue = computationResults.putIfAbsent(key, resultAsync);
+            if (computedValue != null) {
+                return computedValue;
+            }
+            valueLoader.apply(key).convert().toCompletionStage()
+                    .whenComplete((newValue, ex2) -> {
+                        if (ex2 != null) {
+                            resultAsync.completeExceptionally(ex2);
+                            computationResults.remove(key);
+                        } else {
+                            cache.putIfAbsentAsync(key, encodeNull(newValue), lifespan, TimeUnit.MILLISECONDS, maxIdle,
+                                    TimeUnit.MILLISECONDS).whenComplete((existing, ex3) -> {
+                                        if (ex3 != null) {
+                                            resultAsync.completeExceptionally((Throwable) ex3);
+                                        } else if (existing == null) {
+                                            resultAsync.complete(newValue);
+                                        } else {
+                                            resultAsync.complete(decodeNull(existing));
+                                        }
+                                        computationResults.remove(key);
+                                    });
+                        }
+                    });
+            return resultAsync;
+        })).emitOn(new Executor() {
+            // We need make sure we go back to the original context when the cache value is computed.
+            // Otherwise, we would always emit on the context having computed the value, which could
+            // break the duplicated context isolation.
+            @Override
+            public void execute(Runnable command) {
+                Context ctx = Vertx.currentContext();
+                if (context == null) {
+                    // We didn't capture a context
+                    if (ctx == null) {
+                        // We are not on a context => we can execute immediately.
+                        command.run();
```

Comment:
Is it worth to optimize the case when we aren't on a vert.x context in both cases or are on the same context, that we don't have to call emitOn at all?

Reply:
@wburns this code is very similar to the code in client/server mode and reviewed. I would say, let's leave it as it is for now and try to optimize later if needed ?